src/api.py

```diff
@@ -195,27 +201,52 @@ def _generate_with_retry(image_prompt: str, api_inputs: Dict) -> Any:
             raise SteamshipError(f"Invalid runtime parameterization of plugin: {ve}")
 
         openai_result = _generate_with_retry(image_prompt=prompt, api_inputs=inputs)
-        # logging.info("Retry statistics: " + json.dumps(_generate_with_retry.retry.statistics))
-
-        # Fetch data for images
-        urls = [obj.url for obj in openai_result.data]
-
-        return [
-            Block(url=url, mime_type=MimeTypes.PNG, upload_type=BlockUploadType.URL) for url in urls
-        ]
+        return [obj.url for obj in openai_result.data]
 
     def run(
-        self, request: PluginRequest[RawBlockAndTagPluginInput]
+        self, request: PluginRequest[RawBlockAndTagPluginInputWithPreallocatedBlocks]
     ) -> InvocableResponse[RawBlockAndTagPluginOutput]:
         """Run the image generator against all the text, combined."""
         prompt = " ".join([block.text for block in request.data.blocks if block.text is not None])
         user_id = self.context.user_id if self.context is not None else "testing"
-        generated_blocks = self.generate_with_retry(
-            prompt=prompt, user=user_id, options=request.data.options
+        urls = self.generate_with_retry(prompt=prompt, user=user_id, options=request.data.options)
+
+        usage = []
+        output_blocks = request.data.output_blocks
+        for idx, url in enumerate(urls):
+            resp = requests.get(url)
+            if resp.ok:
+                logging.debug(f"[{idx}] appending to block: {output_blocks[idx].id}")
+                output_blocks[idx].append_stream(resp.content)
+                output_blocks[idx].finish_stream()
+                usage.append(
+                    UsageReport(
+                        operation_type=OperationType.RUN,
+                        operation_unit=OperationUnit.UNITS,
+                        operation_amount=1,  # single image generated
+                        audit_id=f"{user_id}-{hash(prompt)}",  # do we want a better audit id ?
+                    )
```

Comment:
i suppose so. though i'm not sure what the "correct" approach to that should be. should we define our plugin-local `operation_unit` for now ( `f"{model}-{image_size}"` ) ?

Reply:
I think another possibility is to include the resolution in the config (that can't be changed in runtime options) and then make that part of the pricing rows

Reply:
that seems less ideal. image sizes _feel_ like they should be dynamic (in the same way number of images is).

Reply:
Yeah, all the options are pretty gross in one way or other.  Operation Units are an enum in the enigne, so adding one requires a change over there. Could also change the units here to just equal cents, but that then requires coordination with pricing spreadsheet deployment.

Reply:
then i think, right now, forcing hard-coding at instance creation is the least ugly?

this is definitely a FR for runtime options fields in billing. i think i had a proposal for that a long time ago -- may be time to resurrect with an actual use case now. one more free-text-y field that plays into billing and this would become unmanageable.

Reply:
Agree that this is least-bad for now.